### ISSUE TITLE
Inconsistent results when running on the same images

### ISSUE THREAD

**Issue by tmurphy0** [openMVG/openMVG#2030]
I'm running a reconstruction multiple times on the same image set. Sometimes there's no result (empty pointcloud), sometimes it's a poor pointcloud, and sometimes it's a good one. Some things I've tried:
- Fix internal calibration and don't allow BA to adjust it. This helped significantly.
- Use priors and no priors. This made no noticeable difference

The following pointclouds were captured fixing the intrinsics and providing priors. I didn't attach the empty empty pointcloud case.

1. good, poses are a bit off
![openmvg01](https://user-images.githubusercontent.com/86392113/159097956-fca6af73-ad28-4745-8002-db512b2ccbba.png)

2. reconstruction looks ok, but poses are way off (notice the green one at the bottom)
![openmvg02](https://user-images.githubusercontent.com/86392113/159097969-4abfba09-8e3e-4f06-94a0-daea263154a7.png)

3. bad reconstruction
![openmvg3](https://user-images.githubusercontent.com/86392113/159098008-de308c3e-2143-4b56-9300-b58d38cff59b.png)

My questions are:
1. How can I get different results, using the exact same input? Is this due to some randomness in the formulation of the BA?
2. Why are the poses so inconsistent on the same dataset? The priors should be fairly accurate.I also tried increasing the wait but didn't notice the poses be closer to the green points.

Any help on how to get more consistent results would be great. Thanks!

**Comment by AaronLK** [comment]
Not to hijack your issue, but I may be seeing something similar. I'm currently investigating using OpenMVG, and have been running various tests on a few aerial datasets. Because the distortion model is different that what I've used in the past, I've been letting OpenMVG optimize the intrinsics, primarily the distortion and principle point. I have been getting some good global results with view priors (loaded as ECEF):

![image](https://user-images.githubusercontent.com/9395531/159179169-5742ad45-5226-4d06-97bb-71d2e1459866.png)

Natuarlly, the next step is to feed the new intrinsics back in and cease optimizing them. But doing so doesn't reproduce the same results. For me, it seems I need to leave optimization on, which isn't ideal.

A quick search of the source code indicates there is one instance of srand in the l2 translation solver:

[srand](https://github.com/openMVG/openMVG/blob/f4161ef98fa676ed8b615194fae2089919a631d3/src/openMVG/multiview/translation_averaging_solver_l2_chordal.cpp#L68)

so I'm going to comment that out and seed the random number generator with the same value on every run. But, it would hopefully be stable enough to not need this. But I thought I'd mention it because it's a good thing to do when trying to track down repeatable results.

**Comment by AaronLK** [comment]
@tmurphy0 how many matches do you have being used?

I realized I was not using the same translation solver that most of the samples used, and switched to the soft L1, and that seems to be producing some stable results now.

I think that srand should be moved elsewhere and set to a constant instead of the current time, however.

**Comment by pmoulon** [comment]
There is multiple factors that is not helping for reproducibility:
- One of them is threading, so some operations are not performed exactly in the same order. It could happen for BA (ceres) and some Robust Estimation steps.
One solution that I know is to disable threading, but it makes reconstruction slow (set OMP_THREAD env variable to 1)

In your case since you have prior we should certainly be able to try to reduce the discrepancy from one run to another by weighing accordingly the cost error related to the 3D priors compare to the number of residual you have

**Comment by qingzhengzhuma** [comment]
This is mainly because threading in translation average(edge coverage algrithm) , not much to with ransac, ceres may lead to different result, but it's influence is very tiny. Once you disable threading in edge coverage algrithm, you will get almost the same result when running on the same dataset serveral times.

**Comment by tmurphy0** [comment]
@pmoulon Thanks for the suggestions! I tried again single threaded (`OMP_NUM_THREADS=1, OMP_THREAD_LIMIT=1`) and multi threaded (`OMP_NUM_THREADS=8, OMP_THREAD_LIMIT=8`). It seems that without threading, the results are consistently bad, while with threading they are sometimes good (though poses still move quite a bit) and sometimes bad or no result. Here are some examples
1. single threaded
![Screenshot from 2022-03-24 10-48-16](https://user-images.githubusercontent.com/86392113/159979864-72e0b8d2-dcd3-452a-9fa2-946d8282498f.png)
![Screenshot from 2022-03-24 10-49-48](https://user-images.githubusercontent.com/86392113/159979866-eda10633-c0f0-4756-9a7b-fede2797f61e.png)
![Screenshot from 2022-03-24 10-51-02](https://user-images.githubusercontent.com/86392113/159979867-3a38d5b0-72cd-49bf-8dd0-de12df7c9611.png)
2. multi threaded
![Screenshot from 2022-03-24 10-52-14](https://user-images.githubusercontent.com/86392113/159979870-80e67039-8c31-4173-a7b7-d036b5236db6.png)
![Screenshot from 2022-03-24 10-52-46](https://user-images.githubusercontent.com/86392113/159979872-fd273a7c-5e26-4450-8e8c-f96687834d4d.png)
![Screenshot from 2022-03-24 10-54-05](https://user-images.githubusercontent.com/86392113/159979875-95279c0a-2fe0-40bb-ad95-060801dec945.png)

At least single threaded makes it consistent, though not necessarily the result I was hoping for :). How come multi threading actually leads to better results? Is there a way I can get consistently good results with single threading?

@qingzhengzhuma Thanks a lot for the explanation! If I understand correctly, you're saying there are 3 sources of randomness (1) threading in translation average, (2) ransac (3) threading in ceres. The major contributor is (1). Running that single threaded should significantly increase consistency without needed to run everything in a single thread.

**Comment by pmoulon** [comment]
Did you try have more point extracted per images (preset -p HIGH on Compute feature) it could help for stability since some pair would be less close to have few support points.

**Comment by qingzhengzhuma** [comment]
@tmurphy0 yes, you got what i mean, as for ransac, because the random seed is fixed, the selected index should be the same no matter how many times you run.

**Comment by pmoulon** [comment]
@tmurphy0 Did you try to see if the results are more stable:
- If you extract more features (feature preset)
- Or use a more accurate matching method (HNSWL2 rather than cascade hashing)?
